src/stackset-stack.ts

```diff
@@ -12,21 +12,94 @@ import {
   Names,
   Lazy,
   FileAssetPackaging,
+  App,
+  Resource,
+  Annotations,
 } from 'aws-cdk-lib';
+import { CfnBucket, IBucket } from 'aws-cdk-lib/aws-s3';
+import { BucketDeployment, Source } from 'aws-cdk-lib/aws-s3-deployment';
 import { Construct } from 'constructs';
 
+export const fileAssetResourceName = 'StackSetAssetsBucketDeployment';
+
 /**
  * Deployment environment for an AWS StackSet stack.
  *
  * Interoperates with the StackSynthesizer of the parent stack.
  */
 export class StackSetStackSynthesizer extends StackSynthesizer {
+  private readonly assetBucket?: IBucket;
+  private bucketDeployment?: BucketDeployment;
+
+  constructor(assetBucket?: IBucket) {
+    super();
+    this.assetBucket = assetBucket;
+  }
+
   public addFileAsset(_asset: FileAssetSource): FileAssetLocation {
-    throw new Error('StackSets cannot use Assets');
+    if (!this.assetBucket) {
+      throw new Error('An Asset Bucket must be provided to use File Assets');
+    }
+
+    if (!_asset.fileName) {
+      throw new Error('Asset filename is undefined');
+    }
+
+    const outdir = App.of(this.boundStack)?.outdir ?? 'cdk.out';
+    const assetPath = `${outdir}/${_asset.fileName}`;
+
+    if (!this.bucketDeployment) {
+      const parentStack = (this.boundStack as StackSetStack)._getParentStack();
+
+      if (!Resource.isOwnedResource(this.assetBucket)) {
+        Annotations.of(parentStack).addWarning('[WARNING] Bucket Policy Permissions cannot be added to' +
+          ' referenced Bucket. Please make sure your bucket has the correct permissions');
+      }
+
+      const bucketDeployment = new BucketDeployment(
+        parentStack,
+        fileAssetResourceName,
+        {
+          sources: [Source.asset(assetPath)],
+          destinationBucket: this.assetBucket,
+          extract: false,
+          prune: false,
+        },
+      );
+
+      this.bucketDeployment = bucketDeployment;
+
+    } else {
+      this.bucketDeployment.addSource(Source.asset(assetPath));
+    }
+
+    const physicalName = this.physicalNameOfBucket(this.assetBucket);
+
+    const bucketName = physicalName;
+    const assetFileBaseName = path.basename(_asset.fileName);
+    const s3Filename = assetFileBaseName.split('.')[1] + '.zip';
+    const objectKey = `${s3Filename}`;
+    const s3ObjectUrl = `s3://${bucketName}/${objectKey}`;
+    const httpUrl = `https://s3.${bucketName}/${objectKey}`;
+
+    return { bucketName, objectKey, httpUrl, s3ObjectUrl };
+  }
+
+  private physicalNameOfBucket(bucket: IBucket) {
+    let resolvedName;
+    if (Resource.isOwnedResource(bucket)) {
+      resolvedName = Stack.of(bucket).resolve((bucket.node.defaultChild as CfnBucket).bucketName);
```

Comment:
the service catalog L2 had this but i'm not sure why it is needed vs just using bucket.bucketName. this part fails an integration test i had when creating a bucket and passing to the construct vs doing a bucket import. it wasn't able to get the name from the token - but this logic works in an actual cdk deploy.